File: kimi-k3/reference/python/kimi_k3_ref/situ.py
"""SiTU-GLU activation matching ``SituAndMul`` in modeling_kimi_linear.py."""

from __future__ import annotations

import numpy as np


def situ_and_mul(
    x: np.ndarray,
    beta: float = 4.0,
    linear_beta: float | None = 25.0,
) -> np.ndarray:
    """
    Official SituAndMul:
        d = x.shape[-1] // 2
        gate = x[..., :d]
        up = x[..., d:]
        situ_a = beta * tanh(gate / beta) * sigmoid(gate)
        if linear_beta is not None:
            up = linear_beta * tanh(up / linear_beta)
        return situ_a * up

    Kimi K3 config: activation_situ_beta=4.0, activation_situ_linear_beta=25.0
    """
    x = np.asarray(x, dtype=np.float32)
    d = x.shape[-1] // 2
    gate = x[..., :d]
    up = x[..., d:]
    beta_f = np.float32(beta)
    situ_a = beta_f * np.tanh(gate / beta_f) * (1.0 / (1.0 + np.exp(-gate)))
    if linear_beta is not None:
        lb = np.float32(linear_beta)
        up = lb * np.tanh(up / lb)
    return (situ_a * up).astype(np.float32)

File: kimi-k3/rust/src/situ.rs
/// SiTU-GLU matching `SituAndMul`.
///
/// `x` is flattened `(..., last_dim)` with even `last_dim`; `out` is `(..., last_dim/2)`.
pub fn situ_and_mul(x: &[f32], last_dim: usize, beta: f32, linear_beta: Option<f32>, out: &mut [f32]) {
    assert!(last_dim % 2 == 0, "SiTU expects even last dim");
    let d = last_dim / 2;
    assert_eq!(x.len() % last_dim, 0);
    assert_eq!(out.len(), x.len() / 2);
    let rows = x.len() / last_dim;
    for r in 0..rows {
        let xs = &x[r * last_dim..(r + 1) * last_dim];
        let os = &mut out[r * d..(r + 1) * d];
        for i in 0..d {
            let gate = xs[i];
            let mut up = xs[d + i];
            let situ_a = beta * (gate / beta).tanh() * sigmoid(gate);
            if let Some(lb) = linear_beta {
                up = lb * (up / lb).tanh();
            }
            os[i] = situ_a * up;
        }
    }
}

#[inline]
fn sigmoid(x: f32) -> f32 {
    1.0 / (1.0 + (-x).exp())
}
